swap/cheque.go

```diff
@@ -0,0 +1,119 @@
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+)
+
+// Encode encodes the cheque in the format used in the signing procedure
+func (cheque *Cheque) Encode() []byte {
+	serialBytes := make([]byte, 32)
+	amountBytes := make([]byte, 32)
+	timeoutBytes := make([]byte, 32)
+	// we need to write the last 8 bytes as we write a uint64 into a 32-byte array
+	// encoded in BigEndian because EVM uses BigEndian encoding
+	binary.BigEndian.PutUint64(serialBytes[24:], cheque.Serial)
+	binary.BigEndian.PutUint64(amountBytes[24:], cheque.Amount)
+	binary.BigEndian.PutUint64(timeoutBytes[24:], cheque.Timeout)
+	// construct the actual cheque
+	input := cheque.Contract.Bytes()
+	input = append(input, cheque.Beneficiary.Bytes()...)
+	input = append(input, serialBytes[:]...)
+	input = append(input, amountBytes[:]...)
+	input = append(input, timeoutBytes[:]...)
+
+	return input
+}
+
+// sigHash hashes the cheque using the prefix that would be added by eth_Sign
+func (cheque *Cheque) sigHash() []byte {
+	input := crypto.Keccak256(cheque.Encode())
+	withPrefix := fmt.Sprintf("\x19Ethereum Signed Message:\n%d%s", len(input), input)
+	return crypto.Keccak256([]byte(withPrefix))
+}
+
+// VerifySig verifies the signature on the cheque
+func (cheque *Cheque) VerifySig(expectedSigner common.Address) error {
+	sigHash := cheque.sigHash()
+
+	if cheque.Sig == nil {
+		return fmt.Errorf("tried to verify signature on cheque with sig nil")
+	}
+	// copy signature to avoid modifying the original
+	sig := make([]byte, len(cheque.Sig))
+	copy(sig, cheque.Sig)
+	// reduce the v value of the signature by 27 (see signContent)
+	sig[len(sig)-1] -= 27
+	pubKey, err := crypto.SigToPub(sigHash, sig)
+	if err != nil {
+		return err
+	}
+
+	if crypto.PubkeyToAddress(*pubKey) != expectedSigner {
+		return ErrInvalidChequeSignature
+	}
+
+	return nil
+}
+
+// Sign signs the cheque with supplied private key
+func (cheque *Cheque) Sign(prv *ecdsa.PrivateKey) ([]byte, error) {
+	sig, err := crypto.Sign(cheque.sigHash(), prv)
+	if err != nil {
+		return nil, err
+	}
+	// increase the v value by 27 as crypto.Sign produces 0 or 1 but the contract only accepts 27 or 28
+	// this is to prevent malleable signatures. while not strictly necessary in this case the ECDSA implementation from Openzeppelin expects it.
+	sig[len(sig)-1] += 27
+	return sig, nil
+}
+
+// Equal checks if other has the same fields
+func (cheque *Cheque) Equal(other *Cheque) bool {
```

Comment:
Do we really need `Equal`? Is comparing just the `Sig` not enough? The sig supposedly is a hashed signed version of the cheque, so if something was different, the `Sig` would be different no?

Reply:
Here we want to verify everything because we use this in tests where we want to make sure that all fields were stored / updated correctly.